### PR TITLE
Pattern support

### LINKED ISSUE #11: Enhance path parameters (Wildcard pattern, relative paths)
In some cases it would be useful to be able to specify a path as:
- relative path
- with wildcard patterns included e.g. war-v*.jar (if a version is included in the jar file)

### PR DESCRIPTION
**Pull request by DreierF**
- [x] [No new Teamscale findings](https://build.cqse.eu/teamscale/delta.html#input/jacoco-client/?showMergeFindings=true)
- [x] [Changes are tested adequately](https://build.cqse.eu/teamscale/tests.html#/jacoco-client/)
- [x] USAGE.md updated in case of user-visible changes
- [x] CHANGELOG.md updated
- [x] build.gradle version number increased according to semantic versioning

Fixes #11

### CODE REVIEW

**Comment by DreierF** [comment]
I have currently no idea why the test is failing on the server. Works fine locally.

File: agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java [hunk 1]
```diff
@@ -76,4 +105,49 @@ public void testHttpServerOptions() throws AgentOptionParseException {
...
+
+		assertFails(".", "**.jar", "Multiple files match the given pattern! " +
+				"Only one match is allowed. Candidates are: [plugins/some_other_file.jar, " +
+				"plugins/file_with_manifest2.jar, file_with_manifest1.jar]");
+
+		assertFails(".", "**.war", "Invalid path given for option option-name: " +
+				"**.war! The pattern **.war did not match any files in");
+	}
+
+	private void assertMatches(String workingDir, String input, String expected) throws AgentOptionParseException, IOException {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java]: Valid point! You easily forget about this if it looks like this in the IDE 😅
image

File: agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java [hunk 2]
```diff
@@ -76,4 +105,49 @@ public void testHttpServerOptions() throws AgentOptionParseException {
...
+		setWorkingDirectory(new File(testFolder.getRoot(), workingDir));
+		Path actualPath = AgentOptionsParser.parsePath("option-name", input);
+		Path expectedPath = new File(testFolder.getRoot(), expected).toPath();
+		assertThat(getCanonicalPath(actualPath)).isEqualByComparingTo(getCanonicalPath(expectedPath));
+	}
+
+	/** Resolves the path to its canonical path. */
+	private static Path getCanonicalPath(Path path) throws IOException {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java]: Your right, normalize should be sufficient

File: agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java [hunk 3]
```diff
@@ -76,4 +105,49 @@ public void testHttpServerOptions() throws AgentOptionParseException {
...
+	}
+
+	private void assertFails(String workingDir, String input, String expectedMessage) {
+		setWorkingDirectory(new File(testFolder.getRoot(), workingDir));
+		assertThatThrownBy(() -> AgentOptionsParser.parsePath("option-name", input))
+				.isInstanceOf(AgentOptionParseException.class).hasMessageContaining(expectedMessage);
+	}
+
+	private static void setWorkingDirectory(File workingDirectory) {
+		System.setProperty("user.dir", workingDirectory.getAbsolutePath());
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java]: Done

**Comment by DreierF** [comment]
rfr

**Comment by karottenreibe** [comment]
The build is failing and done of my comments are bit addressed yet. Did you forget to push some commits maybe?

File: agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java [hunk 4]
```diff
@@ -217,19 +227,20 @@ private static boolean handleTeamscaleOptions(AgentOptions options, String key,
...
-			Manifest manifest = jarStream.getManifest();
-			String branch = manifest.getMainAttributes().getValue("Branch");
-			String timestamp = manifest.getMainAttributes().getValue("Timestamp");
-			if (StringUtils.isEmpty(branch)) {
+	private static CommitDescriptor getCommitFromManifest(Path jarFile) throws AgentOptionParseException {
+		try (JarInputStream jarStream = new JarInputStream(new FileInputStream(jarFile.toFile()))) {
+			Manifest mf = jarStream.getManifest();
+			String branch = mf.getMainAttributes().getValue("Branch");
+			String timestamp = mf.getMainAttributes().getValue("Timestamp");
+			if (branch == null) {
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java]: I didn't remove the isEmpty check and also did not rename the variable to mf 🤔
Totally confused about how this happened

File: agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java [hunk 5]
```diff
@@ -217,19 +227,20 @@ private static boolean handleTeamscaleOptions(AgentOptions options, String key,
...
 				throw new AgentOptionParseException("No entry 'Branch' in MANIFEST!");
 			} else if (StringUtils.isEmpty(timestamp)) {
 				throw new AgentOptionParseException("No entry 'Timestamp' in MANIFEST!");
 			}
+			System.out.println("Found " + branch + ":" + timestamp);
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java]: I added a logger to the AgentOptionsParser which buffers the output and logs it once the actual log4j logger is initialized. I guess we will run into this issue in the future again and would like to keep the logging logic tied to where it actually produces the output .

File: agent/README.md [hunk 6]
```diff
@@ -91,6 +91,11 @@ __Please check the produced log file for errors and warnings before using the ag
 
 The log file is written to the working directory of the profiled Java process by default.
 
+### Path format
+
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/README.md]: Correct me if I'm wrong, but `out` also uses `parsePath(...)` to resolve the path so theoretically a path with pattern could be supplied as well (as long as the directory) already exists.

File: agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java [hunk 7]
```diff
@@ -279,12 +290,80 @@ private static boolean handleHttpServerOptions(AgentOptions options, String key,
...
+
+		return path;
+	}
+
+	/**
+	 * Splits the path into a base dir, a the directory-prefix of the path that does not contain any ? or *
+	 * placeholders, and a pattern suffix.
+	 */
+	private static Pair<String, String> splitIntoBaseDirAndPattern(String value) {
+		String[] pathSegments = value.split("[\\/]");
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java]: Yeah, sounds like a good idea

File: agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java [hunk 8]
```diff
@@ -1,18 +1,46 @@
...
+		testFolder.newFile("file_with_manifest1.jar");
+		testFolder.newFolder("plugins");
+		testFolder.newFolder("plugins", "inner");
+		testFolder.newFile("plugins/some_other_file.jar");
+		testFolder.newFile("plugins/file_with_manifest2.jar");
+	}
+
+	@After
+	public void tearDown() {
+		System.setProperty("user.dir", oldWorkingDir);
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java]: No longer needed

File: agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java [hunk 9]
```diff
@@ -76,4 +105,49 @@ public void testHttpServerOptions() throws AgentOptionParseException {
 		return string -> agentOptions.getLocationIncludeFilter().test(string);
 	}
 
+	/** Tests path resolution with and without patterns and with relative and absolute paths. */
+	@Test
+	public void testPathResolution() throws AgentOptionParseException, IOException {
+		// Test with absolute path
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/test/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsTest.java]: Done

**Comment by DreierF** [comment]
rfr

**Comment by DreierF** [comment]
Tests pass now as well. Was probably an issue with the splitPath method

File: agent/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (karottenreibe) [reply agent/README.md]: sorry, I must have overlooked that

**Comment by karottenreibe** [comment]
see my comments

File: agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java [hunk 10]
```diff
@@ -324,38 +336,37 @@ static Path parsePath(String optionName, File workingDirectory, String value) th
...
 		Pair<String, String> fileAndPattern = splitIntoBaseDirAndPattern(value);
 		String baseDir = fileAndPattern.getFirst();
 		String pattern = fileAndPattern.getSecond();
 
 		File workingDir = workingDirectory.getAbsoluteFile();
-		Path basePath = workingDir.toPath().resolve(baseDir).toAbsolutePath();
+		Path basePath = workingDir.toPath().resolve(baseDir).normalize().toAbsolutePath();
 
 		Pattern pathMatcher = AntPatternUtils.convertPattern(pattern, false);
 		Predicate<Path> filter = path -> pathMatcher.matcher(basePath.relativize(path).toString()).matches();
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java]: Done. Sorry.

File: agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java [hunk 11]
```diff
@@ -324,38 +336,37 @@ static Path parsePath(String optionName, File workingDirectory, String value) th
...
 
 		List<Path> matchingPaths;
 		try {
-			matchingPaths = Files.walk(basePath).filter(filter)
-					.collect(toList());
+			matchingPaths = Files.walk(basePath).filter(filter).sorted().collect(toList());
 		} catch (IOException e) {
 			throw new AgentOptionParseException(
-					"Invalid path given for option " + optionName + ": " + value + "!", e);
+					"Invalid path given for option " + optionName + ": " + value, e);
```
[the remark this answers is not in the document]
Comment (DreierF) [reply agent/src/main/java/eu/cqse/teamscale/jacoco/agent/AgentOptionsParser.java]: Good point!

**Comment by DreierF** [comment]
RFR

File: agent/src/main/java/eu/cqse/teamscale/jacoco/agent/DelayedLogger.java [hunk 12]
```diff
@@ -14,54 +13,49 @@
...
 
 	/** Buffered list of log levels. */
-	private final List<Level> logLevels = new ArrayList<>();
-
-	/** Buffered list of log messages. */
-	private final List<String> logMessages = new ArrayList<>();
-
-	/** Buffered list of logged {@link Throwable}s. */
-	private final List<Throwable> logThrowables = new ArrayList<>();
+	private final List<ILoggerAction> logActions = new ArrayList<>();
```
Comment (karottenreibe) [inline agent/src/main/java/eu/cqse/teamscale/jacoco/agent/DelayedLogger.java]: comment is now outdated

**Comment by karottenreibe** [comment]
One more minor thing. Can you also please do a quick manual test of the path resolution on your machine, just to make sure this works 100% before we ship it? Thanks!

**Comment by DreierF** [comment]
RFR
Tested on Windows and fixed an issue with the *,? characters. With the fix it works

**Comment by karottenreibe** [comment]
Good catch!

**Pull request merged**